pkg/operator/reports.go

```diff
@@ -235,46 +233,85 @@ func (op *Reporting) runReport(logger log.FieldLogger, report *cbTypes.Report) e
-			return nil
+			logger.Infof("report %s is already finished: %s", report.Name, runningCond.Message)
+			return true
 		}
 	}
 
+	return false
+}
+
+// validateReport takes a Report structure and checks if it contains valid fields
+func validateReport(logger log.FieldLogger, report *cbTypes.Report, queryGetter reporting.ReportGenerationQueryGetter, reportDataSourceGetter reporting.ReportDataSourceGetter, reportGetter reporting.ReportGetter) error {
 	// Validate the ReportGenerationQuery is set
 	if report.Spec.GenerationQueryName == "" {
-		return op.setReportStatusInvalidReport(report, "must set spec.generationQuery")
+		return errors.New("must set spec.generationQuery")
 	}
+
 	// Validate the reportingStart and reportingEnd make sense and are set when
-	// required.
+	// required
 	if report.Spec.ReportingStart != nil && report.Spec.ReportingEnd != nil && (report.Spec.ReportingStart.Time.After(report.Spec.ReportingEnd.Time) || report.Spec.ReportingStart.Time.Equal(report.Spec.ReportingEnd.Time)) {
-		return op.setReportStatusInvalidReport(report, fmt.Sprintf("spec.reportingEnd (%s) must be after spec.reportingStart (%s)", report.Spec.ReportingEnd.Time, report.Spec.ReportingStart.Time))
+		return fmt.Errorf("spec.reportingEnd (%s) must be after spec.reportingStart (%s)", report.Spec.ReportingEnd.Time, report.Spec.ReportingStart.Time)
 	}
 	if report.Spec.ReportingEnd == nil && report.Spec.RunImmediately {
-		return op.setReportStatusInvalidReport(report, "spec.reportingEnd must be set if report.spec.runImmediately is true")
+		return errors.New("spec.reportingEnd must be set if report.spec.runImmediately is true")
 	}
 
-	// Validate the ReportGenerationQuery used exists
-	genQuery, err := op.getReportGenerationQueryForReport(report)
+	// Validate the ReportGenerationQuery that the Report used exists
+	genQuery, err := GetReportGenerationQueryForReport(report, queryGetter)
 	if err != nil {
-		logger.WithError(err).Errorf("failed to get report generation query")
 		if apierrors.IsNotFound(err) {
-			return op.setReportStatusInvalidReport(report, fmt.Sprintf("ReportGenerationQuery %s does not exist", report.Spec.GenerationQueryName))
+			return fmt.Errorf("ReportGenerationQuery %s does not exist", report.Spec.GenerationQueryName)
 		}
-		return err
+		return fmt.Errorf("failed to get report generation query")
 	}
 
 	// Validate the dependencies of this Report's query exist
-	queryDependencies, err := reporting.GetAndValidateGenerationQueryDependencies(
-		reporting.NewReportGenerationQueryListerGetter(op.reportGenerationQueryLister),
-		reporting.NewReportDataSourceListerGetter(op.reportDataSourceLister),
-		reporting.NewReportListerGetter(op.reportLister),
+	_, err = reporting.GetAndValidateGenerationQueryDependencies(
+		queryGetter,
+		reportDataSourceGetter,
+		reportGetter,
 		genQuery,
-		op.uninitialiedDependendenciesHandler(),
+		&reporting.UninitialiedDependendenciesHandler{},
```

Comment:
You can pass in the handler as an argument of type `reporting.UninitialiedDependendenciesHandler`